cdap-ui/app/features/hydratorplusplus/controllers/create/toppanel-ctrl.js

```diff
@@ -92,6 +124,116 @@ class HydratorPlusPlusTopPanelCtrl{
+
+      let minutes = Math.floor(duration / 60);
+      let seconds = Math.floor(duration % 60);
+      seconds = seconds < 10 ? '0' + seconds : seconds;
+      minutes = minutes < 10 ? '0' + minutes : minutes;
+
+      this.displayDuration = {
+        minutes: minutes,
+        seconds: seconds
+      };
+    }, 500);
+  }
+  stopTimer() {
+    this.$interval.cancel(this.previewTimerInterval);
+  }
+
+  runPreview() {
+    let startTime = new Date();
+    this.previewStartTime = startTime;
+    this.startTimer();
+    this.previewStore.dispatch(
+      this.previewActions.setPreviewStartTime(startTime)
+    );
+
+    let params = {
+      namespace: this.$state.params.namespace,
+      scope: this.$scope
+    };
+
+
+    // GENERATING PREVIEW CONFIG
+    // This might/should be extracted out to a factory
+
+    let pipelineConfig = this.HydratorPlusPlusConfigStore.getConfigForExport();
+    /**
+     *  This is a cheat way for generating preview for the entire pipeline
+     **/
+
+    let previewConfig = {
+      startStages: [],
+      endStages: [],
+      // useSinks: [], // we are not using sinks for now
+      numOfRecords: 10
+    };
+
+    // Get start stages and end stages
+    // Current implementation:
+    //    - start stages mean sources
+    //    - end stages mean sinks
+    angular.forEach(pipelineConfig.config.stages, (node) => {
+      if (this.GLOBALS.pluginConvert[node.plugin.type] === 'source') {
+        previewConfig.startStages.push(node.name);
+      } else if (this.GLOBALS.pluginConvert[node.plugin.type] === 'sink') {
+        previewConfig.endStages.push(node.name);
+      }
+    });
+    pipelineConfig.config.preview = previewConfig;
+
+    if (previewConfig.startStages.length === 0 || previewConfig.endStages.length === 0) {
+      this.HydratorPlusPlusConsoleActions.addMessage({
+        type: 'danger',
+        content: 'Please add a source and sink to the pipeline'
+      });
+      return;
+    }
+
+    this.myPipelineApi.runPreview(params, pipelineConfig).$promise
+      .then((res) => {
+        this.previewStore.dispatch(
+          this.previewActions.setPreviewId(res.preview)
+        );
+
+        this.startPollPreviewStatus(res.preview);
+      }, (err) => {
+        this.stopTimer();
+        this.HydratorPlusPlusConsoleActions.addMessage({
+          type: 'danger',
+          content: err.data
+        });
+      });
+  }
+
+  startPollPreviewStatus(previewId) {
+    this.dataSrc.poll({
+      _cdapNsPath: '/previews/' + previewId + '/status',
+      interval: 5000
+    }, (res) => {
+      if (res.status !== 'RUNNING') {
+        // this.stopTimer();
+        // this.dataSrc.stopPoll(res.__pollId__);
+
+
+        /* THIS IS ONLY FOR DEMO. REMOVE and uncomment the above */
+        setTimeout(() => {
+          this.stopTimer();
+          this.dataSrc.stopPoll(res.__pollId__);
+        }, 5000);
+      }
+    });
+  }
+
+  togglePreviewMode() {
+    this.previewStore.dispatch(
```

Comment:
talked with @ajainarayanan, we are not going to stop the polling for now until the backend adds the stop preview endpoint